src/main.rs

```diff
@@ -35,20 +41,63 @@ fn main() -> Result<(), Error> {
     );
 
     let opt = Opt::from_args();
-    let config: Config = toml::from_str(&fs::read_to_string(opt.config)?)?;
+    let mut config: Config = toml::from_str(&fs::read_to_string(opt.config)?)?;
+
+    match config.validate() {
+        Ok(_) => (),
+        Err(e) => return Err(Error::ValidationError(e)),
+    };
+    for camera in &config.cameras {
+        match camera.validate() {
+            Ok(_) => (),
+            Err(e) => return Err(Error::ValidationError(e)),
+        };
+    }
+
+    // Setup auto sub streams, we do this by looping the cameras
+    // If the stream type is both we clone the config
+    // On one of the clones we set stream_source to "mainStream" on the other "subStream"
+    // We also change the mount name by appending "/mainStream" or "/subStream"
+    // On the original uncloned config we change the stream from "both" to "mainStream" and leave the mount name unchanged
+    let mut new_cam_configs = vec![];
+
+    for camera_config in &mut config.cameras {
+        if camera_config.stream == "both" {
+            let mut main_camera_config = camera_config.clone();
+            let mut sub_camera_config = camera_config.clone();
+
+            camera_config.stream = "mainStream".to_string();
+
+            main_camera_config.stream = "mainStream".to_string();
+            main_camera_config.name = format!("{}/{}", main_camera_config.name, main_camera_config.stream);
+            new_cam_configs.push(main_camera_config);
+
+            sub_camera_config.stream = "subStream".to_string();
+            sub_camera_config.name = format!("{}/{}", sub_camera_config.name, sub_camera_config.stream);
+            sub_camera_config.format = "h264".to_string(); // Assuming always H264 on subStream: TODO: Autodetect
+            new_cam_configs.push(sub_camera_config);
+        }
```

Comment:
This approach will cause the main stream to stream from the camera twice, once for each RTSP path `camera/` and `camera/mainStream`. A better approach would be to call `add_stream()` once for each stream, possibly giving it more than one path. The "trick" here would be that you can call `crossbeam::scope::spawn` as many times as you like, not just once per loop.

If you want me to do this particular change, I can show you what I mean.

Reply:
This means we would need to supply more than one path in the config and the loop over `mounts.add_factory(&format!("/{}", name), &factory);` with each path right? Is there a way to add a path array to the config structure without it being avaliable via toml I wonder... Hmm time to read more docs.

Reply:
Aha! this `#[serde(skip)]` :)

Reply:
Could use skip, although you don't need to change the config structure to get the code to do what you want here, I don't think. I'll write a patch on your branch if you don't mind, and go ahead and merge as it is now.

Reply:
I see well it should be ok to merge then. I did try what I suggested with multiple paths and looping `mounts.add_factory` but it fails to setup the mounts. The docs at [https://gstreamer.freedesktop.org/documentation/gst-rtsp-server/rtsp-mount-points.html?gi-language=c#GstRTSPMountPoints] seems to suggest that once called the factory should not be used again and that calling add_factory also removes any previous mounts. Perhaps I am going about this the wrong way.

Reply:
So I have just updated the readme and validation and will now stop, as you say lets add the functionality and maybe make another PR for improvements when time permits. Thanks!

Reply:
Great! `mounts.add_factory` should be ok to call more than once, because the docs are talking about consuming the pointer you pass in, not the underlying Factory, which can be shared. This is transparent in the Rust bindings. (I *think*. I am not a Gstreamer expert.)